File: .github/workflow-scripts/src/update-readme.js
import fs from 'fs/promises';
import { getDeployAddresses, saveDeployInfo, loadDeployInfo } from './utils/deploy-info.js';
import { generateComponentHashes } from './utils/component.js';
import { COMPONENTS, COMPONENT_TYPES } from './workflow-config.js';
import { getClient, queryAllBalances } from './utils/chain.js';

async function generateReadmeData() {
    // Get environment variables
    const {
        GIST_SECRET,
        GIST_ID,
        DEPLOYER_ADDRESS,
        MINTER_ADDRESS,
        OWNER_ADDRESS,
        USER_ADDRESS
    } = process.env;

    if (!GIST_SECRET || !GIST_ID) {
        throw new Error('Missing required environment variables');
    }

    // Calculate component hashes
    console.log('Calculating component hashes...');
    const hashes = generateComponentHashes(COMPONENTS);

    // Get latest deploy info and balances
    console.log('Getting latest deploy info and balances...');
    const [addresses, client] = await Promise.all([
        getDeployAddresses(GIST_ID, GIST_SECRET),
        getClient()
    ]);

    const addressMap = {
        deployer: DEPLOYER_ADDRESS,
        minter: MINTER_ADDRESS,
        owner: OWNER_ADDRESS,
        user: USER_ADDRESS
    };

    const balances = await queryAllBalances(client, addressMap);

    // Prepare data
    const data = {
        lastUpdated: new Date().toISOString(),
        hashes: {
            frontend: hashes[COMPONENT_TYPES.FRONTEND].slice(0, 8),
            mosaicTile: hashes[COMPONENT_TYPES.MOSAIC_TILE].slice(0, 8),
        },
        deploy: {
            timestamp: addresses.timestamp,
            mosaicTileAddress: addresses.minter,
        },
        balances: Object.entries(addressMap).reduce((acc, [role, address]) => ({
            ...acc,
            [role]: {
                address,
                balance: balances[role]
            }
        }), {})
    };

    return data;
}

async function updateReadme(data) {
    // Read template
    console.log('Reading template...');
    const template = await fs.readFile('README.template.md', 'utf8');

    // Replace template variables
    let output = template
        .replace('{{ .LastUpdated }}', data.lastUpdated)
        .replace('{{ .Hashes.Frontend }}', data.hashes.frontend)
        .replace('{{ .Hashes.MosaicTile }}', data.hashes.mosaicTile)
        .replace('{{ .Deploy.Timestamp }}', data.deploy.timestamp)
        .replace('{{ .Deploy.MosaicTileAddress }}', data.deploy.mosaicTileAddress)

    // Replace balance section
    const balanceRows = Object.entries(data.balances)
        .map(([role, data]) => `| ${role} | \`${data.address}\` | ${data.balance} |`)
        .join('\n');

    output = output.replace(/\{\{- range \$role, \$data := \.Balances \}\}[\s\S]*?\{\{- end \}\}/, balanceRows);

    // Write output
    console.log('Writing README...');
    await fs.writeFile('README.md', output);
    console.log('README.md updated successfully');
}

async function main() {
    try {
        // Generate data
        const data = await generateReadmeData();

        // Save data to JSON file
        await saveDeployInfo(data);

        // Update README with the data
        await updateReadme(data);

    } catch (error) {
        console.error('Error:', error);
        process.exit(1);
    }
}

main(); 

File: .github/workflow-scripts/update-readme.go
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"net/http"
	"os"
	"path/filepath"
	"text/template"
	"time"
)

type ComponentHashes struct {
	Frontend      string
	MosaicTile    string
}

type DeployInfo struct {
	Timestamp          string
	MosaicTileAddress string
}

type BalanceInfo struct {
	Address string
	Balance string
}

type TemplateData struct {
	LastUpdated string
	Hashes      ComponentHashes
	Deploy      DeployInfo
	Balances    map[string]BalanceInfo
}

func calculateDirectoryHash(path string) (string, error) {
	hash := sha256.New()
	err := filepath.Walk(path, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if !info.IsDir() {
			data, err := ioutil.ReadFile(path)
			if err != nil {
				return err
			}
			hash.Write(data)
		}
		return nil
	})
	if err != nil {
		return "", err
	}
	return hex.EncodeToString(hash.Sum(nil))[:8], nil
}

func getLatestDeployInfo() (DeployInfo, error) {
	gistId := os.Getenv("GIST_ID")
	token := os.Getenv("GITHUB_TOKEN")
	url := fmt.Sprintf("https://api.github.com/gists/%s", gistId)

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return DeployInfo{}, err
	}

	req.Header.Set("Authorization", "token "+token)
	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return DeployInfo{}, err
	}
	defer resp.Body.Close()

	var gist struct {
		Files map[string]struct {
			Content string `json:"content"`
		} `json:"files"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&gist); err != nil {
		return DeployInfo{}, err
	}

	var latestDeploy struct {
		Timestamp string `json:"timestamp"`
		Job struct {
			Data struct {
				MosaicTileAddress    string `json:"mosaic_tile_address"`
			} `json:"data"`
		} `json:"job"`
	}

	var deployInfo DeployInfo
	for filename, file := range gist.Files {
		if filename == "mosaic_tile_nft_deploy.json" {
			if err := json.Unmarshal([]byte(file.Content), &latestDeploy); err != nil {
				continue
			}
			deployInfo = DeployInfo{
				Timestamp:            latestDeploy.Timestamp,
				MosaicTileAddress:    latestDeploy.Job.Data.MosaicTileAddress,
			}
			break
		}
	}

	return deployInfo, nil
}

func main() {
	// Calculate component hashes
	frontendHash, err := calculateDirectoryHash("frontend")
	if err != nil {
		fmt.Printf("Error calculating frontend hash: %v\n", err)
		os.Exit(1)
	}

	mosaicTileHash, err := calculateDirectoryHash("contracts/mosaic_tile_nft")
	if err != nil {
		fmt.Printf("Error calculating mosaic tile hash: %v\n", err)
		os.Exit(1)
	}


	// Get latest deploy info
	deployInfo, err := getLatestDeployInfo()
	if err != nil {
		fmt.Printf("Error getting deploy info: %v\n", err)
		os.Exit(1)
	}

	// Prepare template data
	data := TemplateData{
		LastUpdated: time.Now().UTC().Format(time.RFC3339),
		Hashes: ComponentHashes{
			Frontend:      frontendHash,
			MosaicTile:    mosaicTileHash,
		},
		Deploy: deployInfo,
		Balances: map[string]BalanceInfo{
			"deployer": {Address: os.Getenv("DEPLOYER_ADDRESS")},
			"minter":   {Address: os.Getenv("MINTER_ADDRESS")},
			"owner":    {Address: os.Getenv("OWNER_ADDRESS")},
			"user":     {Address: os.Getenv("USER_ADDRESS")},
		},
	}

	// Read template
	tmpl, err := template.ParseFiles("README.template.md")
	if err != nil {
		fmt.Printf("Error parsing template: %v\n", err)
		os.Exit(1)
	}

	// Create output file
	output, err := os.Create("README.md")
	if err != nil {
		fmt.Printf("Error creating output file: %v\n", err)
		os.Exit(1)
	}
	defer output.Close()

	// Execute template
	if err := tmpl.Execute(output, data); err != nil {
		fmt.Printf("Error executing template: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("README.md updated successfully")
} 